### PR TITLE
[GlobalOpt] Re-enable transpose propagation

### PR DESCRIPTION
**Pull request by qedawkins**
Based on the regression in #16055, this re-enabled transpose propagation by default and also defers matmul_transpose_b/a raising to the propagation pass.

### CODE REVIEW

**Comment by hanhanW** [comment]
The regressions on CPU side are all on no-dt path. The CPU default logic assumes that all the contraction ops are in `nk.km->nm` form. However now we have more transpose_b ops. Perhaps the fix is to interchange the distribution/cache/vector tile sizes.

image

**Comment by MaheshRavishankar** [comment]
> The regressions on CPU side are all on no-dt path. The CPU default logic assumes that all the contraction ops are in `nk.km->nm` form. However now we have more transpose_b ops. Perhaps the fix is to interchange the distribution/cache/vector tile sizes.
> 
> image

Interesting. Yeah probably a good point.

**Comment by hanhanW** [comment]
I was wrong about interchanging. The tile sizes are still set on corresponding dimension. The issue is that we are generating a `vector.transpose` op when lowering `vector.contract`; the new `vector.transpose` op is not optimized. It becomes

```mlir
        %8 = vector.transpose %7, [1, 0] : vector<32x16xf32> to vector<16x32xf32>
        %9 = vector.extract %8[0] : vector<32xf32> from vector<16x32xf32>
        %10 = memref.load %subview[%arg0, %arg2] : memref<32x128xf32, strided<[128, 1], offset: ?>, #hal.descriptor_type<storage_buffer>>
        %11 = vector.broadcast %10 : f32 to vector<32xf32>
        %12 = vector.extract %arg3[0] : vector<32xf32> from vector<8x32xf32>
        %13 = vector.fma %11, %9, %12 : vector<32xf32>
        %14 = affine.apply affine_map<()[s0] -> (s0 + 1)>()[%arg0]
        %15 = memref.load %subview[%14, %arg2] : memref<32x128xf32, strided<[128, 1], offset: ?>, #hal.descriptor_type<storage_buffer>>
        %16 = vector.broadcast %15 : f32 to vector<32xf32>
        %17 = vector.extract %arg3[1] : vector<32xf32> from vector<8x32xf32>
        %18 = vector.fma %16, %9, %17 : vector<32xf32>
...
```

Without raising it to `linalg.matmul_transpose_b`, CPU backend generates good transpose code. But here it falls back to 1D shuffle, which is bad. I'm trying to break it down to two 16x16 transpose, and let's see if it helps...

**Comment by hanhanW** [comment]
https://github.com/openxla/iree/pull/16485#issuecomment-1953361965 shows that it is 2x faster (520 ms -> 263 ms for MobileBert_fp16), but it does not address all the performance issue. The original perf number of MobileBert_fp16 is 180 ms. I will take another deep look tomorrow.

**Comment by hanhanW** [comment]
Without enabling the transpose propagation, the `transpose + matmul` op is not converted into `matmul_transpose_b` op. 

In `matmul_transpos_b` case, the `8, 32, 16` tile sizes are selected. Thus, it is lowered to below snippet, which introduces a transposition on RHS with 32x16 shape.

```mlir
%6 = vector.transpose %5, [1, 0] : vector<32x16xf32> to vector<16x32xf32>
// .. the rest of IRs is as same as linalg.matmul lowering
```

The 32x16 2D transpose is lowered to a big 1-D shuffle, which is not efficient.

Without raising the op to `matmul_transpose_b`, the x86 backend can generate good code for both `transpose` and `matmul` dispatches. We generate a sequence of `vunpcklps+vunpckhps` and `vshuff64x2` ops for transpose, while it is not the case in `matmul_transpose_b`.

Now I have a prototype which can generates good transpose code for `matmul_transpose_b` case. The transpose part is lowered to a 2-trip loop, which apply 16x16 transpose separately and store them back to a stack buffer. This allows us to generate the sequence of `vunpcklps+vunpckhps` and `vshuff64x2` ops. Here is the [asm](https://gist.githubusercontent.com/hanhanW/51c0dcdfd244e5c8395fd46e1a0fe81b/raw/a0588db3ea27e909d3105997feeecdd833e32e72/out.asm).

| transpose+matmul | matmul_transpose_b w/o opt | matmul_transpose_b w/ opt |
|------------------|----------------------------|---------------------------|
|         0.405 ms |                     1.3 ms |                  0.519 ms |

We can put the transpose code and matmul code together in `matmul_transpose_b` code, but we still have performance gap. IMO, the difference come from memory access. The solutions for non-DT seem to be very tricky. This class of issues are well covered by data-tiling. In this context, do we still have action items for non-DT path? Is it a blocker to land the PR?

Side note: this also impacts ARM workloads, e.g., DeepLabV3, etc. It is only happening on non-DT path as well.

image

 @dcaballe we'd like to enable transpose propagation by default. This is impacting non-DT matmuls because of `matmul_transpose_b` codegen. For x86 side, I narrowed the issue down to memory accessing. I did not get a chance on ARM side. I wonder if the regressions on non-DT path would become a blocker? If so, what's your suggestions to move this forward?

**Comment by dcaballe** [comment]
It should be ok as we don't have a good matmul story on the non-DT path at this point but I would make sure this is not interfering with @banach-space work.

**Comment by MaheshRavishankar** [comment]
Thanks @hanhanW for the deep dive here! I think the small regression you are seeing in non-DT path seems managable to me. I am assuming that this is based on a local change in your branch and not reflected in the benchmark numbers yet?

**Comment by banach-space** [comment]
> I did not get a chance on ARM side.

I’d offer to help, but am OOO.

I am also worried about the SVE and SME lowerings. In particular, for the latter you’d transpose LHS rather than RHS. So there’s 2 more cases that will be affected, but are not investigated. But since we are not tracking SVE nor SME perf upstream, this should not be a blocker.

Would using “-iree-global-opt-propagate-transposes=false” preserve the current behaviour? This way we’d have an obvious workaround for SVE and SME. That would be preferred.

**Comment by qedawkins** [comment]
> > I did not get a chance on ARM side.
> 
> I’d offer to help, but am OOO.
> 
> I am also worried about the SVE and SME lowerings. In particular, for the latter you’d transpose LHS rather than RHS. So there’s 2 more cases that will be affected, but are not investigated. But since we are not tracking SVE nor SME perf upstream, this should not be a blocker.
> 
> Would using “-iree-global-opt-propagate-transposes=false” preserve the current behavior? This way we’d have an obvious workaround for SVE and SME. That would be preferred.

Yes, `-iree-global-opt-propagate-transposes=false` would preserve the current behaviour. I'll add that in an ideal world, this pass is orthogonal to what the preferred layout of a matrix multiplication is. Data tiling (i.e. `linalg_ext.set_encoding` -> `tensor.pack`) is the mechanism for setting preferred layouts for a particular target. SetEncodings runs immediately after this pass and is where we would expect to introduce transposes on the `LHS` and/or `RHS` (or do data tiling) based on the needs of SVE/SME (although the transposes shouldn't be materialized until codegen).

This pass is just to try to fuse some of the common extra transposes we see on the inputs of matmuls (among other things) coming from frontends, and thus give a more canonical input to SetEncodings. We can think of this pass as more of a canonicalization pass rather than trying to take an opinion on the layout of a matrix. Of course there are only so many engineering hours so we have flags in the interim :)

**Comment by benvanik** [comment]
Caught my interest: are we talking about global transposes on the LHS of matmuls being required on SVE? If so, that's going to be real bad for users multitargeting. Or is this just the tiles processed by a workgroup? That'd be fine(ish).

**Comment by banach-space** [comment]
>  are we talking about global transposes on the LHS of matmuls being required on SVE?

That should be similar to what we do elsewhere, so I guess local? (sorry, I’ve nicotine paid much attention to multi-target yet). 

Btw, this would be required for SME (Scalable _Matrix_ Extension) rather than SVE (Scalable _Vector_ Extension). The former computes matmul via outer products (rather than dot-product), hence transposing LHS rather than RHS.

Tl;Dr We shouldn’t require anything special here. Well,  apart from transposing LHS instead of RHS.

> Yes, -iree-global-opt-propagate-transposes=false would preserve the current behaviour.

👍🏻  Would it be possible to verify with a test? Or is that already tested?

**Comment by MaheshRavishankar** [comment]
> > are we talking about global transposes on the LHS of matmuls being required on SVE?
> 
> That should be similar to what we do elsewhere, so I guess local? (sorry, I’ve nicotine paid much attention to multi-target yet).
> 
> Btw, this would be required for SME (Scalable _Matrix_ Extension) rather than SVE (Scalable _Vector_ Extension). The former computes matmul via outer products (rather than dot-product), hence transposing LHS rather than RHS.
> 
> Tl;Dr We shouldn’t require anything special here. Well, apart from transposing LHS instead of RHS.
> 
> > Yes, -iree-global-opt-propagate-transposes=false would preserve the current behaviour.
> 
> 👍🏻 Would it be possible to verify with a test? Or is that already tested?

You should be really using data tiling to do this. You can then pick the inner tile sizes appropraitetly for SME or SVE

**Comment by banach-space** [comment]
> You should be really using data tiling to do this. You can then pick the inner tile sizes appropraitetly for SME or SVE

Sure, but from what I can tell, we can’t lower linalg.mmt4d with masking enabled (let alone with scalable sizes). One missing piece implemented here:
* https://github.com/llvm/llvm-project/pull/81906#pullrequestreview-1891200219

Also, doesn’t mmt4d assume that it’s RHS that’s transposed? And we need LHS transposed instead. In any way, many of these things boil down to:
> Of course there are only so many engineering hours

:)

**Comment by qedawkins** [comment]
Closing this in favor of #17238
